### ISSUE TITLE
Inconsistent false positive "inconsistent overloaded implementation" when the return value is a value that pyright infers as literal

### ISSUE THREAD

**Issue by Wizzerinus** [microsoft/pyright#9676]
**Describe the bug**
The code below causes Pyright to say that "int is not assignable to a literal". (This example may look artificial but the same happens in a lot of cases, and I don't fully understand when it is triggered and when it is not triggered.)

**Code or Screenshots**
This is one example when this false positive happens:

```python
from typing import overload, Literal

@overload
def foo(n: Literal[False]) -> int: ...

@overload
def foo(n: Literal[True]) -> int: ...

def foo(n: bool):
    if n:
        return 1
    else:
        return 0
```

But in this case there is no error:
```python
from typing import overload, Literal

def bar() -> int: ...  # to stop pyright from inferring the type as a Literal

@overload
def foo(n: Literal[False]) -> int: ...

@overload
def foo(n: Literal[True]) -> int: ...

def foo(n: bool):
    if n:
        return bar()
    else:
        return 0
```

But in this case it happens again (only for the tuple overload):
```python
from typing import overload, Literal

def bar() -> int: ...

@overload
def foo(n: Literal[False]) -> int: ...

@overload
def foo(n: Literal[True]) -> tuple[int, int]: ...

def foo(n: bool):
    if n:
        return 0, bar()
    else:
        return bar()
```

**VS Code extension or command-line**
CLI, version 1.1.391.

**Comment by erictraut** [comment]
Pyright is working as designed here. You haven't provided a return type annotation for the implementation of `foo`, so pyright uses return type inference to infer the type based on return statements within the function. In this case, the inferred return type is `Literal[0, 1]`. You can fix this by simply providing a return type annotation of `-> int`.
